### PR TITLE
shims: add a shim for gethostname

### LINKED ISSUE #5138: Missing support for `gethostname`
Now that some Tokio-dependent test can run, I discovered that `gethostname` is called in some networking code, which I think should be fairly straightforward to add.

```
test behavior::tests::test_async_handler_works_with_pending_internal_future ... error: unsupported operation: can't call foreign function `gethostname` on OS `linux`
   --> /home/nazar-pc/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/resolv-conf-0.7.6/src/lib.rs:553:20
    |
553 |                 if gethostname(hostname.as_mut_ptr(), hostname.len()) < 0 {
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unsupported operation occurred here
    |
    = help: this means the program tried to do something Miri does not support; it does not indicate a bug in the program
    = note: this is on thread `behavior::tests`
    = note: stack backtrace:
            0: resolv_conf::Config::get_system_domain
                at /home/nazar-pc/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/resolv-conf-0.7.6/src/lib.rs:553:20: 553:70
            1: hickory_resolver::system_conf::unix::into_resolver_config
                at /home/nazar-pc/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hickory-resolver-0.26.1/src/system_conf/unix.rs:44:40: 44:73
            2: hickory_resolver::system_conf::unix::parse_resolv_conf::<std::vec::Vec<u8>>
                at /home/nazar-pc/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hickory-resolver-0.26.1/src/system_conf/unix.rs:37:5: 37:38
            3: hickory_resolver::system_conf::unix::read_resolv_conf::<&str>
                at /home/nazar-pc/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hickory-resolver-0.26.1/src/system_conf/unix.rs:29:5: 29:39
            4: hickory_resolver::system_conf::unix::read_system_conf
                at /home/nazar-pc/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hickory-resolver-0.26.1/src/system_conf/unix.rs:25:5: 25:41
...
```

### PR DESCRIPTION
**Pull request by faukah**
Add a shim for gethostname().
Closes rust-lang/miri#5138.

Behavior differs between targets, so I had to match on Os *and* on Glibc/Musl. I don't know whether `throw_unsup_format!()` is the correct way of handling all other cases. 

I'm not *quite* sure how Solaris behaves, but I assume it matches the behavior of Illumos. Also not quite sure if MacOS behavior is correct. It *can* in theory return `ENAMETOOLONG`, but docs and source clash a bit here:

Manpage: 

>  [ENAMETOOLONG]     The current host name is longer than namelen.  (For gethostname() only.)

Source code:
```c
int
gethostname(char *name, size_t namelen)
{
	int mib[2];

	mib[0] = CTL_KERN;
	mib[1] = KERN_HOSTNAME;
	if (namelen < MAXHOSTNAMELEN + 1) {
		char local_buf[MAXHOSTNAMELEN + 1];
		size_t local_len = sizeof(local_buf);
		if (sysctl(mib, 2, local_buf, &local_len, NULL, 0) == -1) {
			if (errno == ENOMEM)
				errno = ENAMETOOLONG;
			return (-1);
		}
		strncpy(name, local_buf, namelen);
		name[namelen - 1] = 0;
	} else {
	if (sysctl(mib, 2, name, &namelen, NULL, 0) == -1) {
		if (errno == ENOMEM)
			errno = ENAMETOOLONG;
		return (-1);
		}
	}
	return (0);
}
```
As far as I can see, if `sysctl()` succeeds, i.e. the hostname is copied into `local_buf`, and `name` is too small, it still would put a NULL-terminated part of the hostname into `name` and return successfully.       

A few notes on the `match` statement:

@ Android

```rust
(Os::Android, _) => this.set_errno_and_return_neg1_i32(LibcError("ENAMETOOLONG")),
```

Android returns error code `ENAMETOOLONG` without writing anything to `name`, see [here](https://raw.githubusercontent.com/aosp-mirror/platform_bionic/master/libc/bionic/gethostname.cpp):

```c
int gethostname(char* buf, size_t n) {
  utsname name = {};
  uname(&name);

  size_t name_length = static_cast<size_t>(strlen(name.nodename) + 1);
  if (name_length > n) {
    errno = ENAMETOOLONG;
    return -1;
  }

  memcpy(buf, name.nodename, name_length);
  return 0;
}
```

@ GNU libc, FreeBSD

GNU libc and FreeBSD behave the same, see [libc source](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/posix/gethostname.c;h=b17671373ca1ac61e7346ae0dcdda22eff1d7fcd;hb=HEAD) and [FreeBSD source](https://github.com/freebsd/freebsd-src/blob/main/lib/libc/gen/gethostname.c). Both GNU libc and FreeBSD write a trunctuated, _non-NULL-terminated_ prefix of the hostname before returning:

```rust
(Os::Linux, Env::Gnu) | (Os::FreeBsd, _) => {
    let len: usize = len.try_into().unwrap();
    this.write_bytes_ptr(name, hostname[..len].iter().copied())?;
    this.set_errno_and_return_neg1_i32(LibcError("ENAMETOOLONG"))
}
```

GNU libc:

```c
node_len = strlen (buf.nodename) + 1;
memcpy (name, buf.nodename, len < node_len ? len : node_len);

if (node_len > len)
  {
    __set_errno (ENAMETOOLONG);
    return -1;
  }
```

FreeBSD:

```c
int mib[2];

mib[0] = CTL_KERN;
mib[1] = KERN_HOSTNAME;
if (sysctl(mib, 2, name, &namelen, NULL, 0) == -1) {
	if (errno == ENOMEM)
		errno = ENAMETOOLONG;
	return (-1);
}
```

@ Musl, MacOS, Solaris, Illumos

```rust
(Os::Linux, _) | (Os::MacOs, _) | (Os::Solaris, _) | (Os::Illumos, _) => {
    if len > 0 {
        let len: usize = len.try_into().unwrap();
        this.write_bytes_ptr(
            name,
            hostname[..len - 1].iter().copied().chain(std::iter::once(0)),
        )?;
    }
    interp_ok(Scalar::from_i32(0))
}
```

[Musl libc](https://git.musl-libc.org/cgit/musl/tree/src/unistd/gethostname.c)'s implementation is actually different from GNU libc, it adds a NULL byte at the end:

```c
if (len > sizeof uts.nodename) len = sizeof uts.nodename;
for (i=0; i<len && (name[i] = uts.nodename[i]); i++);
if (i && i==len) name[i-1] = 0;
}
```

In cases where hostname length == buffer length, we get different behavior:

```sh
$ ./gethostname_musl.o
ret=0 errno=0 (No error information)
bytes=74 65 6d 00
chars=tem\0

$ ./gethostname_glibc.o
ret=-1 errno=36 (File name too long)
bytes=74 65 6d 70
chars=temp
```

MacOS [also NULL-terminates the name](https://github.com/apple-open-source/macos/blob/master/Libc/gen/FreeBSD/gethostname.c):

```c
char local_buf[MAXHOSTNAMELEN + 1];
size_t local_len = sizeof(local_buf);
if (sysctl(mib, 2, local_buf, &local_len, NULL, 0) == -1) {
	if (errno == ENOMEM)
		errno = ENAMETOOLONG;
	return (-1);
}
strncpy(name, local_buf, namelen);
name[namelen - 1] = 0;
```

Illumos [uses the systeminfo syscall](https://raw.githubusercontent.com/illumos/illumos-gate/master/usr/src/lib/libc/port/gen/gethostname.c):

```c
return (sysinfo(SI_HOSTNAME, name, namelen) == -1 ? -1 : 0);
```

And systeminfo also NULL-terminates the buffer when it's too small:

```c
if (kstr != NULL) {
		strcnt = strlen(kstr);
		if (count > 0) {
			if (count <= strcnt) {
				getcnt = count - 1;
				if (subyte(buf + getcnt, 0) < 0)
					return (set_errno(EFAULT));
			} else {
				getcnt = strcnt + 1;
			}
			if (copyout(kstr, buf, getcnt))
				return (set_errno(EFAULT));
		}
		return (strcnt + 1);
	}
```

Relevant manpages:
- MacOS: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/gethostname.3.html
- Linux: https://www.man7.org/linux/man-pages/man2/gethostname.2.html
- FreeBSD: https://man.freebsd.org/cgi/man.cgi?query=gethostname
- Solaris: https://docs.oracle.com/cd/E88353_01/html/E37843/gethostname-3c.html

### CODE REVIEW

File: src/shims/unix/foreign_items.rs [hunk 1]
```diff
@@ -164,6 +164,21 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 let result = this.getcwd(buf, size)?;
                 this.write_pointer(result, dest)?;
             }
+            "gethostname" => {
+                this.check_target_os(
+                    &[Os::Linux, Os::Android, Os::MacOs, Os::FreeBsd, Os::Solaris, Os::Illumos],
```
Comment (RalfJung) [inline src/shims/unix/foreign_items.rs]: Isn't that all the OSes we support? Doesn't seem necessary to have such a check then (unless we already know that there are other Unixes where this would be wrong).

File: src/shims/unix/env.rs [hunk 2]
```diff
@@ -219,6 +219,52 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+        this.assert_target_os_is_unix("gethostname");
+
+        let name = this.read_pointer(name_op)?;
+        let len = this.read_target_usize(len_op)?;
+
+        if this.ptr_is_null(name)? {
+            return this.set_errno_and_return_neg1_i32(LibcError("EFAULT"));
+        }
+
+        let hostname = b"Miri";
```
Comment (RalfJung) [inline src/shims/unix/env.rs]: Seems worth making this a global constant so it's not hidden in the middle of this function. 

You can add it next to this:
https://github.com/rust-lang/miri/blob/1a4be0157a648b8bef565e5d7c2a57838f449b69/src/shims/unix/mod.rs#L33

File: src/shims/unix/env.rs [hunk 3]
```diff
@@ -219,6 +219,52 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+        if len > u64::try_from(hostname.len()).unwrap() {
+            let (written, _) = this.write_c_str(hostname, name, len)?;
+            assert!(written); // Value should fit.
+            return interp_ok(Scalar::from_i32(0));
+        }
+
+        match (&this.tcx.sess.target.os, &this.tcx.sess.target.env) {
```
Comment (RalfJung) [inline src/shims/unix/env.rs:262]:
```suggestion
        // The exact behavior on a too short bugger differs by platform and even libc.
        // POSIX says:
        // > The returned name shall be null-terminated, except that if namelen is
        // > an insufficient length to hold the host name, then the returned name
        // > shall be truncated and it is unspecified whether the returned name is
        // > null-terminated.
        match (&this.tcx.sess.target.os, &this.tcx.sess.target.env) {
```

File: src/shims/unix/env.rs [hunk 4]
```diff
@@ -219,6 +219,52 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+            (Os::Android, _) => this.set_errno_and_return_neg1_i32(LibcError("ENAMETOOLONG")),
+            (Os::Linux, Env::Gnu) | (Os::FreeBsd, _) => {
+                let len: usize = len.try_into().unwrap();
```
Comment (RalfJung) [inline src/shims/unix/env.rs:270]:
```suggestion
                // Write what we can *without* trailing null.
                let len: usize = len.try_into().unwrap();
```

File: src/shims/unix/env.rs [hunk 5]
```diff
@@ -219,6 +219,52 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+                this.write_bytes_ptr(name, hostname[..len].iter().copied())?;
+                this.set_errno_and_return_neg1_i32(LibcError("ENAMETOOLONG"))
+            }
+            (Os::Linux, _) | (Os::MacOs, _) | (Os::Solaris, _) | (Os::Illumos, _) => {
+                if len > 0 {
+                    let len: usize = len.try_into().unwrap();
+                    let copy_len = len.checked_sub(1).unwrap();
```
Comment (RalfJung) [inline src/shims/unix/env.rs]:
```suggestion
                    let copy_len = len.strict_sub(1);
```

File: src/shims/unix/env.rs [hunk 6]
```diff
@@ -219,6 +219,52 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+
+        match (&this.tcx.sess.target.os, &this.tcx.sess.target.env) {
+            (Os::Android, _) => this.set_errno_and_return_neg1_i32(LibcError("ENAMETOOLONG")),
+            (Os::Linux, Env::Gnu) | (Os::FreeBsd, _) => {
+                let len: usize = len.try_into().unwrap();
+                this.write_bytes_ptr(name, hostname[..len].iter().copied())?;
+                this.set_errno_and_return_neg1_i32(LibcError("ENAMETOOLONG"))
+            }
+            (Os::Linux, _) | (Os::MacOs, _) | (Os::Solaris, _) | (Os::Illumos, _) => {
+                if len > 0 {
```
Comment (RalfJung) [inline src/shims/unix/env.rs]:
```suggestion
                // Write what we can *with* trailing null.
                if len > 0 {
```

File: src/shims/unix/env.rs [hunk 7]
```diff
@@ -219,6 +219,52 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+
+        let hostname = b"Miri";
+        if len > u64::try_from(hostname.len()).unwrap() {
+            let (written, _) = this.write_c_str(hostname, name, len)?;
+            assert!(written); // Value should fit.
+            return interp_ok(Scalar::from_i32(0));
+        }
+
+        match (&this.tcx.sess.target.os, &this.tcx.sess.target.env) {
+            (Os::Android, _) => this.set_errno_and_return_neg1_i32(LibcError("ENAMETOOLONG")),
```
Comment (RalfJung) [inline src/shims/unix/env.rs]:
```suggestion
            (Os::Android, _) => {
                // Android violates POSIX and does not write anything:
                // <https://raw.githubusercontent.com/aosp-mirror/platform_bionic/master/libc/bionic/gethostname.cpp>
                this.set_errno_and_return_neg1_i32(LibcError("ENAMETOOLONG"))
            }
```

File: src/shims/unix/env.rs [hunk 8]
```diff
@@ -219,6 +219,52 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+            (Os::Linux, _) | (Os::MacOs, _) | (Os::Solaris, _) | (Os::Illumos, _) => {
+                if len > 0 {
+                    let len: usize = len.try_into().unwrap();
+                    let copy_len = len.checked_sub(1).unwrap();
+                    this.write_bytes_ptr(
+                        name,
+                        hostname[..copy_len].iter().copied().chain(std::iter::once(0)),
+                    )?;
+                }
+                interp_ok(Scalar::from_i32(0))
```
Comment (RalfJung) [inline src/shims/unix/env.rs:290]: Why does this return 0? POSIX says
> Upon successful completion, 0 shall be returned; otherwise, -1 shall be returned.

So we should have some sort of comment with a link for every one of these targets showing that they do indeed violate POSIX.

File: tests/pass-dep/libc/libc-gethostname.rs [hunk 9]
```diff
@@ -0,0 +1,58 @@
+//@ignore-target: windows # No libc
```
Comment (RalfJung) [inline tests/pass-dep/libc/libc-gethostname.rs:1]:
```suggestion
//@ignore-target: windows # No libc
//@run-native
```
We only run these on Linux (glibc) and macOS, but better than nothing.

File: tests/pass-dep/libc/libc-gethostname.rs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+fn test_too_small() {
+    let mut name = [0u8; 4];
+    let result = unsafe { libc::gethostname(name.as_mut_ptr().cast(), name.len()) };
+    cfg_select! {
+        target_os = "android" => {
+            let err = errno_result(result).unwrap_err();
+            assert_eq!(&name, &[0u8; 4]);
+            assert_eq!(err.raw_os_error(), Some(libc::ENAMETOOLONG));
+            assert_eq!(io::Error::last_os_error().raw_os_error(), Some(libc::ENAMETOOLONG));
```
Comment (RalfJung) [inline tests/pass-dep/libc/libc-gethostname.rs]:
```suggestion
```
This is redundant.

File: tests/pass-dep/libc/libc-gethostname.rs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        }
+        any(all(target_os = "linux", target_env = "gnu"), target_os = "freebsd") => {
+            let err = errno_result(result).unwrap_err();
+            assert_eq!(&name, b"Miri");
+            assert_eq!(err.raw_os_error(), Some(libc::ENAMETOOLONG));
+            assert_eq!(io::Error::last_os_error().raw_os_error(), Some(libc::ENAMETOOLONG));
```
Comment (RalfJung) [inline tests/pass-dep/libc/libc-gethostname.rs]:
```suggestion
```

File: tests/pass-dep/libc/libc-gethostname.rs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        _ => {
+            compile_error!("unsupported target");
+        }
+    }
+}
+
+fn test_null_ptr() {
+    let err = errno_result(unsafe { libc::gethostname(ptr::null_mut(), 5) }).unwrap_err();
+    assert_eq!(err.raw_os_error(), Some(libc::EFAULT));
+    assert_eq!(io::Error::last_os_error().raw_os_error(), Some(libc::EFAULT));
```
Comment (RalfJung) [inline tests/pass-dep/libc/libc-gethostname.rs]:
```suggestion
```

**Review by RalfJung — COMMENTED** [review]
Sorry for the slow review, and thanks for all the research! Impressive how inconsistent platforms apparently are...

@rustbot author

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5145/e063ce5d5a9cb4ac6323952754af0c8b348b1eac..f13c9bccf8efae763093a25fca03612c0ec6bbed)*

File: src/shims/unix/env.rs (hunk 8 above)
Comment (faukah) [reply src/shims/unix/env.rs:290]: Because it completes successfully. POSIX also says

> The returned name shall be null-terminated, except that if namelen is an insufficient length to hold the host name, then the returned name shall be truncated and it is unspecified whether the returned name is null-terminated.

It says nowhere that this is an error, and Musl, MacOS, Solaris, and Illumos all match this behavior.

File: src/shims/unix/env.rs (hunk 3 above)
Comment (faukah) [reply src/shims/unix/env.rs:262]: I assume this is supposed to mean *buffer*.

File: src/shims/unix/env.rs (hunk 8 above)
Comment (RalfJung) [reply src/shims/unix/env.rs:290]: Sounds like an error to me...
Please explicitly explain this in a comment, and add a link for musl (and maybe more links if you want).

File: src/shims/unix/env.rs (hunk 3 above)
Comment (RalfJung) [reply src/shims/unix/env.rs:262]: Yes :)

File: tests/pass-dep/libc/libc-gethostname.rs (hunk 9 above)
Comment (RalfJung) [reply tests/pass-dep/libc/libc-gethostname.rs:1]: Ah right the test will need adjustment because when run natively it sees the actual hostname of the GH CI runner.

Comment (faukah) [reply tests/pass-dep/libc/libc-gethostname.rs:1]: Yup. I'm not *quite* sure on how to best resolve this. Something like 
```rust
fn main() {
    let hostname = test_ok();
    test_too_small(&hostname);

    test_null_ptr();
}
```
would probably work, but I don't really want tests to rely on each other.

Comment (RalfJung) [reply tests/pass-dep/libc/libc-gethostname.rs:1]: That seems fine to me for a case like this.

Comment (faukah) [reply tests/pass-dep/libc/libc-gethostname.rs:1]: Something like this could work, I guess?
```diff
diff --git a/tests/pass-dep/libc/libc-gethostname.rs b/tests/pass-dep/libc/libc-gethostname.rs
index f6deac989a..9d6016b461 100644
--- a/tests/pass-dep/libc/libc-gethostname.rs
+++ b/tests/pass-dep/libc/libc-gethostname.rs
@@ -5,34 +5,49 @@
 mod libc_utils;
 
 use std::ffi::CStr;
-use std::ptr;
 
 use libc_utils::*;
 
 fn main() {
     test_ok();
     test_too_small();
+
+    #[cfg(miri)]
     test_null_ptr();
 }
 
+fn get_hostname() -> Vec<u8> {
+    let mut name = [0u8; 1024];
+    errno_check(unsafe { libc::gethostname(name.as_mut_ptr().cast(), name.len()) });
+
+    let hostname = unsafe { CStr::from_ptr(name.as_ptr().cast()) }.to_bytes().to_owned();
+    assert!(!hostname.is_empty());
+    hostname
+}
+
 fn test_ok() {
-    let mut name = [0u8; 5];
-    errno_check(unsafe { libc::gethostname(name.as_mut_ptr().cast(), name.len()) });
-    assert_eq!(unsafe { CStr::from_ptr(name.as_ptr().cast()) }, c"Miri");
+    let hostname = get_hostname();
+
+    if cfg!(miri) {
+        assert_eq!(hostname, b"Miri");
+    }
 }
 
 fn test_too_small() {
-    let mut name = [0u8; 4];
+    let hostname = get_hostname();
+
+    // Leave no room for the trailing null byte.
+    let mut name = vec![0u8; hostname.len()];
     let result = unsafe { libc::gethostname(name.as_mut_ptr().cast(), name.len()) };
     cfg_select! {
         target_os = "android" => {
             let err = errno_result(result).unwrap_err();
-            assert_eq!(&name, &[0u8; 4]);
+            assert!(name.iter().all(|byte| *byte == 0));
             assert_eq!(err.raw_os_error(), Some(libc::ENAMETOOLONG));
         }
         any(all(target_os = "linux", target_env = "gnu"), target_os = "freebsd") => {
             let err = errno_result(result).unwrap_err();
-            assert_eq!(&name, b"Miri");
+            assert_eq!(name, hostname);
             assert_eq!(err.raw_os_error(), Some(libc::ENAMETOOLONG));
         }
         any(
@@ -42,7 +57,9 @@
             target_os = "solaris",
         ) => {
             errno_check(result);
-            assert_eq!(&name, b"Mir\0");
+            let mut expected = hostname.to_owned();
+            *expected.last_mut().unwrap() = 0;
+            assert_eq!(name, expected);
         }
         _ => {
             compile_error!("unsupported target");
@@ -50,7 +67,10 @@
     }
 }
 
+// Some native libcs dereference `name` in userspace and may crash on null.
+// This test covers Miri's explicit `EFAULT` handling only.
+#[cfg(miri)]
 fn test_null_ptr() {
-    let err = errno_result(unsafe { libc::gethostname(ptr::null_mut(), 5) }).unwrap_err();
+    let err = errno_result(unsafe { libc::gethostname(std::ptr::null_mut(), 5) }).unwrap_err();
     assert_eq!(err.raw_os_error(), Some(libc::EFAULT));
 }
```

Not quite sure if there's a nicer way of doing it.

Comment (faukah) [reply tests/pass-dep/libc/libc-gethostname.rs:1]: > That seems fine to me for a case like this.

GitHub hid that comment until now, I'll implement that then. I'm uncertain about how to handle `test_null_ptr()` best though, that fails in userspace and SIGSEVs when running natively. I suppose the best thing would be to add a `#[cfg(miri)]` directive?

Comment (RalfJung) [reply tests/pass-dep/libc/libc-gethostname.rs:1]: > Some native libcs dereference `name` in userspace and may crash on null.

If that's a thing then Miri should report UB for such calls.
That should happen naturally if you just remove the null check in the implementation.

File: tests/fail-dep/libc/libc-gethostname-null.rs [hunk 13]
```diff
@@ -0,0 +1,7 @@
+//@ignore-target: windows # No libc
+
+fn main() {
+    unsafe {
+        libc::gethostname(std::ptr::null_mut(), 5); //~ERROR: null pointer
```
Comment (RalfJung) [inline tests/fail-dep/libc/libc-gethostname-null.rs:8]: Please add a comment explaining that some libc deref this pointer (ideally with a link).

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5145/e063ce5d5a9cb4ac6323952754af0c8b348b1eac..5d2ff39fbc280ef449721fe75a2bce78ca700fac)*

File: tests/pass-dep/libc/libc-gethostname.rs [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+use libc_utils::*;
+
+fn main() {
+    test_ok();
+
+    test_too_small();
+}
+
+fn get_hostname() -> Vec<u8> {
+    let mut name = [0u8; 1024];
```
Comment (RalfJung) [inline tests/pass-dep/libc/libc-gethostname.rs]:
```suggestion
    let mut name = [0u8; 1024]; // we assume this is enough space
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5145/e063ce5d5a9cb4ac6323952754af0c8b348b1eac..5d2ff39fbc280ef449721fe75a2bce78ca700fac)*

File: tests/pass-dep/libc/libc-gethostname.rs (hunk 9 above)
Comment (faukah) [reply tests/pass-dep/libc/libc-gethostname.rs:1]: I can think of a single edge case left, then. What happens when we call `gethostname(NULL, 0)`? Behavior differs across implementations as well here, e.g. glibc declares `name` as `nonnull`. 

> That should happen naturally

From what I can see, Miri would not catch that case. If we deem this UB as well, we probably need to add an explicit check.

EDIT: this is UB for MacOS: https://github.com/apple-open-source/macos/blob/master/Libc/gen/FreeBSD/gethostname.c#L60
```c
name[namelen - 1] = 0;
```

Comment (RalfJung) [reply tests/pass-dep/libc/libc-gethostname.rs:1]: Ah good point. Yeah that should be UB.

That macOS version means that even `gethostname(ptr, 0)` is UB if `ptr` points to the end of an allocation, doesn't it?

Comment (RalfJung) [reply tests/pass-dep/libc/libc-gethostname.rs:1]: Actually no that macOS code makes a size of 0 always UB no matter the `ptr`. This is an array access at index `usize::MAX`! I can see no way to argue that this is POSIX-compliant...

Still, we should probably do the same then: `throw_ub_format!` if the size is 0.

File: tests/fail-dep/libc/libc-gethostname-zero-length.rs [hunk 15]
```diff
@@ -0,0 +1,11 @@
+//@only-target: apple
+
+fn main() {
+    // MacOS unconditionally writes to `name[namelen - 1]`, so passing zero for `namelen`
+    // causes an out-of-bounds access:
+    // <https://github.com/apple-oss-distributions/Libc/blob/main/gen/FreeBSD/gethostname.c#L48-L60>
```
Comment (faukah) [inline tests/fail-dep/libc/libc-gethostname-zero-length.rs:6]: I'm not a big fan of duplicating this comment, but I think it'd be good to document this here as well.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5145/e063ce5d5a9cb4ac6323952754af0c8b348b1eac..4518f854e65f72dc58ceea0f883caa5a31a18c8f)*

File: src/shims/unix/env.rs [hunk 16]
```diff
@@ -255,6 +255,12 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 this.write_bytes_ptr(name, HOSTNAME[..len].iter().copied())?;
                 this.set_errno_and_return_neg1_i32(LibcError("ENAMETOOLONG"))
             }
+            (Os::MacOs, _) if len == 0 => {
```
Comment (RalfJung) [inline src/shims/unix/env.rs]: I was actually thinking we should just make this UB everywhere. Who knows which horrors lurk in the other implementations.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5145/e063ce5d5a9cb4ac6323952754af0c8b348b1eac..4518f854e65f72dc58ceea0f883caa5a31a18c8f)*

Comment (faukah) [reply src/shims/unix/env.rs]: `len == 0` with a valid buffer seems fine (as in no UB) on all other platforms. 

For those, it probably makes sense to reject `gethostname(NULL, x)` but allow `gethostname(VALID_BUFFER, 0)`

Comment (faukah) [reply src/shims/unix/env.rs]: Fwiw, I think this is probably the "best" behavior.

| Input | Miri output |
|---|---|
| `gethostname(valid_pointer, 0)` | Allowed on all platforms except macOS |
| `gethostname(valid_pointer, 0)` on macOS | UB due to `name[namelen - 1]` |
| `gethostname(NULL, x)` for any `x` | UB on every platform |

Comment (RalfJung) [reply src/shims/unix/env.rs]: I'm not a fan of overfitting, and I am inclined to take what macOS does as a signal that the length must be non-zero. Then we can just add a `check_ptr_access` to ensure that `len` many bytes are available, which will exclude null even on Android.

File: src/shims/unix/env.rs [hunk 17]
```diff
@@ -219,6 +220,77 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+        name_op: &OpTy<'tcx>,
+        len_op: &OpTy<'tcx>,
+    ) -> InterpResult<'tcx, Scalar> {
+        let this = self.eval_context_mut();
+        this.assert_target_os_is_unix("gethostname");
+
+        let name = this.read_pointer(name_op)?;
+        let len = this.read_target_usize(len_op)?;
+
+        // Require the length argument to be non-zero on every platform. This cannot be folded into
```
Comment (RalfJung) [inline src/shims/unix/env.rs]:
```suggestion
        // macOS makes a length of 0 UB by writing to `name[namelen - 1]`:
        // <https://github.com/apple-oss-distributions/Libc/blob/main/gen/FreeBSD/gethostname.c#L48-L60>.
        // Not know what POSIX actually intends, we require the length argument to be non-zero on every platform. This cannot be folded into
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5145/e063ce5d5a9cb4ac6323952754af0c8b348b1eac..ac93c791d72d6abe612e2788781e9e65e8406f64)*

File: src/shims/unix/env.rs [hunk 18]
```diff
@@ -255,12 +266,6 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
             }
-            (Os::MacOs, _) if len == 0 => {
-                // MacOS unconditionally writes to `name[namelen - 1]`, so passing zero for
-                // `namelen` causes an out-of-bounds access:
-                // <https://github.com/apple-oss-distributions/Libc/blob/main/gen/FreeBSD/gethostname.c#L48-L60>
-                throw_ub_format!("`gethostname` called with `namelen` set to zero on macOS")
-            }
             (Os::Linux, _) | (Os::MacOs, _) | (Os::Solaris, _) | (Os::Illumos, _) => {
                 // Write what we can *with* trailing null.
                 if len > 0 {
```
Comment (RalfJung) [inline src/shims/unix/env.rs]: This `if` is unnecessary now.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/5145/e063ce5d5a9cb4ac6323952754af0c8b348b1eac..ac93c791d72d6abe612e2788781e9e65e8406f64)*

**Comment by RalfJung** [comment]
@rustbot author

Also, as we are nearing the end of the review process, please squash the commits. :)

**Comment by faukah** [comment]
squashed into 6664947075913ee2134e95392cf964a5a834d667. 

@rustbot ready

**Comment by RalfJung** [comment]
Thanks a lot :)
